### ISSUE TITLE
Incorporating tf data into the NixOS configuration

### ISSUE THREAD

**Issue by shlevy** [Gabriella439/terraform-nixos-ng#6]
Currently, if you want to automatically incorporate some information from the tf state into your system configuration (e.g. make an SSH key owned by tf that can connect to the machine), you have to either write your flake to hard-code some path that will be generated by tf, or use tf's `local_file` machinery to write a flake to local (presumably gitignored) disk and use that as your flakeref.

I can think of two possible paths to improve this:

1. Make `nixos-rebuild` aware of some kind of command-line configurable variant of a NixOS config at the flakeref. The simplest thing might be to just allow `nixosConfigurations.foo` to be a function which is called using the typical `--arg`/`--argstr` mechanism (maybe augmented with `--argflakeref` :sweat_smile: ).
2. Add some parameters to this tf module to allow for inline specification of a flake with tf interpolation, which under the hood manages the temporary flake file creation for you.

I like 1 more, but it's obviously out of scope for this project and I suspect there will be upstream resistance. Would you be interested in 2 or some other in-scope solution?

**Comment by Gabriella439** [comment]
I'm also partial to (1) myself.  Is the issue that it would require extending `nix build` (for flakes) to support `--arg`/`--argstr`?  Because that's something I've been interested in myself

**Comment by shlevy** [comment]
One idea I had that I may make a quick PR/RFC for: An `inline` flakeref type, where the rest of the "URL" is just parsed as a flake.nix attrset. E.g. I could do `nix build 'inline:{ outputs = { nixpkgs, ... }: { foo = nixpkgs.legacyPackages.x86_64-linux.hello; }; }#foo'` as a really dumb way to build `hello`. Then any configuration can just be a `config` input.

Could even perhaps have `inline:[ true, 1 ]?flake=false` :grimacing:

**Comment by shlevy** [comment]
But yes, the idea I proposed in 1 above would require either having `nix build` support `--arg` or having `nixos-rebuild` build the drv through some other call path than `nix build`

**Comment by shlevy** [comment]
:crossed_fingers:

**Comment by roberth** [comment]
If it has a parameter (such as data from terraform here), it's not a configuration, but a module, so you shouldn't be using `nixosConfigurations`, but rather `nixosModules`.
This might not correspond to our normal vague use of the word configuration, but definitions matter.

**Comment by shlevy** [comment]
If we're being precise (and I agree that that is important), what's suggested here is neither a `nixosConfiguration` nor a `nixosModule` but a `function yielding a nixosConfiguration which may under the hood do so primarily by importing an existing nixosModule`. But as I mentioned in https://github.com/NixOS/nix/pull/8815#issuecomment-1676334823, the mechanism of being able to autocall in flake evaluation is orthogonal to the policy of how one should semantically refer to the values one wants to autocall.

**Comment by roberth** [comment]
> function yielding a nixosConfiguration which may under the hood do so primarily by importing an existing nixosModule

That would be a sufficient solution, but it doesn't leverage modules, which are more declarative than functions.

**Leveraging modules**

Modules are functions, syntactically, and semantically they are functions from a partial configuration to a differently partial configuration. When producing a module from a function parameter (ie not the module arguments), generally the same can be achieved with an option.
When you write a function instead of declaring options, you split the configuration in two: the parameter(s) and the options. This reduces composability.

**Declarative**

I define a declarative configuration system as the ideal where
 - all configuration happens through declarations*, which are statements about the domain that are as independent as possible
 - these declarations compose trivially

Functions interfere with the latter, and generally don't conform to the first.
\*: both declarations and definitions in module system terminology

> orthogonal

That's why we have both this thread and the other. Indeed let's not mix up the discussions.

**Comment by Gabriella439** [comment]
@roberth: It's not clear to me if you are proposing a different solution to this problem

**Comment by roberth** [comment]
Right, I was somewhat suggesting to focus on modules. I'll elaborate on that.

Flake based projects that use terraform-nixos-ng don't have pure configurations in the sense of both the module system and flakes, so they shouldn't use `nixosConfigurations`.
Instead, terraform-nixos-ng could call `nixosSystem` and have parameters to select a module / modules from `nixosModules`, and perhaps `specialArgs` as well.

Ideally it also knows how to generate a typical "hardware-configuration.nix" module, or at least the parts it can know. `nixpkgs.hostPlatform` in particular would be rather redundant information if specified manually (e.g. `system` parameter in `nixosSystem`). The architecture is just a "function" of the instance type. This might be too high level knowledge for a simple terraform module though. It'd probably have to be a provider-specific module to achieve that, but maybe terraform has improved since I last checked.

**Comment by Gabriella439** [comment]
So the reason `terraform-nixos-ng` uses the `nixosConfigurations` attribute is because [that's what `nixos-rebuild` uses](https://github.com/NixOS/nixpkgs/blob/d58f7137e0ed0dd90efebbea0254548bd0babcb6/pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh#L369).  I don't have strong opinions on what the flake output should be named but I want the solution to be `nixos-rebuild` compatible because I don't want to reinvent what `nixos-rebuild` is already doing.

**Comment by roberth** [comment]
NixOS is still a bit underdeveloped in this area as of yet, area being whatever it is that `nixos-rebuild` currently fills. I blame the bystander effect. I know two other tool authors who suffer from this, and I am complicit myself... Also there's a [first-mover disadvantage](https://en.wikipedia.org/wiki/Law_of_the_handicap_of_a_head_start) aspect, where high level NixOS came before low level Flakes.

Anyway, I think it is a responsibility of NixOS to provide a suitable entrypoint; one that's Nixlang-first instead of CLI first, so that it can be integrated into tools more easily.
Also I think there's an opportunity to standardize the core of "module system applications" such as NixOS, where we have certain options (chiefly `toplevel`), a CLI, activation script, profile switching, perhaps ssh, but that doesn't necessarily have to be a part of the NixOS change just yet.
